fix: Adjust tool pattern to support multi-line inputs

vblagoje opened the pull request:
### Related Issues
- fixes https://github.com/deepset-ai/haystack/issues/4575

### Proposed Changes:
 Change the default ToolsManager tools pattern to match multiple lines of tool input yet be backwards compatible with the old tools pattern

### How did you test it?
All the existing tools' pattern tests are run against the new default regex pattern
Added a unit test specifically for the use case of multi-line input `test_extract_tool_name_and_tool_multi_line_input`

### Notes for the reviewer
Please try to find a counter-example that breaks the pattern. 

### Checklist
- [x] I have read the [contributors guidelines](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack/blob/main/code_of_conduct.txt)
- [x] I have updated the related issue with new insights and changes
- [x] I added tests that demonstrate the correct behavior of the change
- [x] I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`.
- [x] I documented my code
- [x] I ran [pre-commit hooks](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md#installation) and fixed any issue

vblagoje commented:
@julian-risch this is still a draft until we merge https://github.com/deepset-ai/haystack/pull/4794

vblagoje commented:
@julian-risch it seems this PR fails for Tool Input that has quotes. Let me have a look at this first, add unit tests etc. I'll ping you when ready

julian-risch reviewed (approved):
Looks good to me! 👍 I tried some examples and couldn't find a problem. The only behavior one could discuss is that an empty string as tool input results in the regex not matching and thus None, None will be returned. In case we want to allow an empty tool input, the + character in the last matching group needs to be changed to *, I think: `(?:.|\n)+)`
What was helpful for testing was the following interpretation of the regex (yes, you know where I have that from):
- \s*: matches zero or more whitespace characters at the beginning of the string.
- (?: ... ): a non-capturing group used to group together two alternative patterns.
- \"([\s\S]*?)\": the first alternative pattern, which matches a double-quoted string. It starts and ends with a double quote (\"), and contains any number of characters, including whitespace and newlines, as captured by the ([\s\S]*?) group. The ? makes the * quantifier lazy, meaning that it matches as few characters as possible while still allowing the entire pattern to match.
- |: the alternation operator, used to separate the two alternative patterns.
- ((?:.|\n)+): the second alternative pattern, which matches any string of one or more characters that is not surrounded by double quotes. It consists of a capturing group ((?:.|\n)+) that matches any character (.) or newline (\n) one or more times. The (?: ... ) is a non-capturing group used to group together the . and \n characters within the + quantifier, which matches one or more occurrences of the group.
- \s*: matches zero or more whitespace characters at the end of the string.

tholor commented:
What's missing?

vblagoje commented:
@julian-risch, thanks for the insights. I have changed the + to * and added more tests. The main branch to which we rebased now has even more tooling tests, so if this CI run is green - let's integrate it.
